src/combinatorics/additive/erdos_ginzburg_ziv.lean

```diff
@@ -0,0 +1,264 @@
+   (filter (λ a, p ↑a) s.attach).card = (filter p s).card :=
+by simp_rw [←countp_eq_card_filter, countp_attach]
+
+end multiset
+
+namespace finset
+variables {α β : Type*} [comm_monoid β] (s : finset α) (f : α → β)
+
+open_locale big_operators
+
+@[simp, to_additive] lemma prod_map_val : (s.1.map f).prod = ∏ a in s, f a := rfl
+
+end finset
+
+namespace finset
+variables {α β : Type*} [add_comm_monoid_with_one β]
+
+open_locale big_operators
+
+@[simp] lemma card_val (s : finset α) : s.1.card = s.card := rfl
+
+lemma card_filter (p : α → Prop) [decidable_pred p] (s : finset α) :
+  (filter p s).card = ∑ a in s, ite (p a) 1 0 :=
+sum_boole.symm
+
+lemma coe_card_filter (p : α → Prop) [decidable_pred p] (s : finset α) :
+  ((filter p s).card : β) = ∑ a in s, ite (p a) 1 0 :=
+by { rw card_filter, norm_cast }
+
+@[simp] lemma card_filter_attach (p : α → Prop) [decidable_pred p] (s : finset α) :
+   (filter (λ a, p ↑a) s.attach).card = (filter p s).card :=
+multiset.card_filter_attach _ _
+
+end finset
+
+namespace finset
+variables {α β : Type*} [semilattice_sup α] [order_bot α] {f : β → α} {s : finset β} {a : α}
+
+lemma sup_const_le : s.sup (λ _, a) ≤ a := finset.sup_le $ λ _ _, le_rfl
+
+end finset
+
+namespace finset
+variables {α β : Type*} [semilattice_inf α] [order_top α] {f : β → α} {s : finset β} {a : α}
+
+lemma le_inf_const_le : a ≤ s.inf (λ _, a) := finset.le_inf $ λ _ _, le_rfl
+
+end finset
+
+namespace nat
+variables {a b : ℕ}
+
+lemma eq_of_dvd_of_lt_two_mul (ha : a ≠ 0) (hdvd : b ∣ a) (hlt : a < 2 * b) : a = b :=
+begin
+  obtain ⟨_ | _ | c, rfl⟩ := hdvd,
+  { simpa using ha },
+  { exact mul_one _ },
+  { cases hlt.not_le ((mul_comm _ _).trans_le $ mul_le_mul_left' (one_lt_succ_succ _) _) }
+end
+
+end nat
+
+namespace mv_polynomial
+variables {R S₁ σ : Type*} [semiring R] [comm_semiring S₁] [module R S₁] {a : R}
+  {f : mv_polynomial σ S₁}
+
+lemma support_smul : (a • f).support ⊆ f.support :=
+λ n, by { simp_rw mem_support_iff, exact right_ne_zero_of_smul }
+
+lemma total_degree_smul_le : (a • f).total_degree ≤ f.total_degree := finset.sup_mono support_smul
+
+@[simp] lemma constant_coeff_smul (a : R) (f : mv_polynomial σ S₁) :
+  constant_coeff (a • f) = a • constant_coeff f := rfl
+
+end mv_polynomial
+
+namespace zmod
+variables {p : ℕ} [fact (nat.prime p)]
+
+lemma pow_card_sub_one (x : zmod p) : x ^ (p - 1) = if x ≠ 0 then 1 else 0 :=
+begin
+  split_ifs with hx,
+  { exact pow_card_sub_one_eq_one hx },
+  { simp [of_not_not hx, (fact.out p.prime).one_lt] }
+end
+
+end zmod
+
+open multiset mv_polynomial
+open_locale big_operators
+
+namespace zmod
+variables {p : ℕ} [fact p.prime] {s : multiset (zmod p)}
+
+/-- The two multivariate polynomials used in the proof of Erdős–Ginzburg–Ziv. -/
+private noncomputable def f (s : multiset (zmod p)) : bool → mv_polynomial s.to_enum_finset (zmod p)
+| ff := ∑ x in s.to_enum_finset.attach, X x ^ (p - 1)
+| tt := ∑ x in s.to_enum_finset.attach, x.1.1 • X x ^ (p - 1)
```

Comment:
I am using an indexed family because that's what [`char_dvd_card_solutions_family`](https://leanprover-community.github.io/mathlib_docs/find/char_dvd_card_solutions_family) expects, but I could instead write a binary version if you think that's better.

Reply:
Binary version in #18083